src/Assignments/ProblemEditor.tsx

```diff
@@ -0,0 +1,391 @@
+
+    const render = () => {
+        setPreviewState({
+            seedValue: problemEditorForm.getValues().seedValue,
+            showHints: problemEditorForm.getValues().showHints,
+            showSolutions: problemEditorForm.getValues().showSolutions,
+            problemSource: problemEditorForm.getValues().problemSource,
+        });
+    };
+
+    const load = async () => {
+        try {
+            const result = await readProblem({
+                filePath: problemEditorForm.getValues().loadPath
+            });
+            problemEditorForm.setValue(nameof<ProblemEditorInputs>('problemSource'), result.data.data.problemSource);
+            const userPath = getSavePathForLoadPath(problemEditorForm.getValues().loadPath);
+            problemEditorForm.setValue(nameof<ProblemEditorInputs>('userPath'), userPath);
+            render();
+        } catch(e) {
+            logger.error('problemEditor: load: ', e.message);
+            setAlertState({
+                message: e.message,
+                variant: 'danger'
+            });
+        }
+    };
+
+    const save = async () => {
+        try {
+            const { problemSource } = problemEditorForm.getValues();
+            if(_.isNil(problemSource)) {
+                return;
+            }
+            const result = await saveProblem({
+                problemSource: problemSource,
+                relativePath: problemEditorForm.getValues().userPath,
+            });
+            const savedPath = result.data.data.filePath;
+            problemEditorForm.setValue(nameof<ProblemEditorInputs>('loadPath'), savedPath);
+            setMyCatalog(myCatalog => [...myCatalog, savedPath]);
+            setAlertState({
+                message: 'Saved successfully',
+                variant: 'success'
+            });            
+            render();
+        } catch(e) {
+            logger.error('problemEditor: save:', e.message);
+            setAlertState({
+                message: e.message,
+                variant: 'danger'
+            });
+        }
+    };
+
+    const getCatalog = async () => {
+        try {
+            const result = await catalog();
+            setMyCatalog([...myCatalog, ...result.data.data.problems]);
+        } catch(e) {
+            logger.error('problemEditor: getCatalog:', e.message);
+            setAlertState({
+                message: e.message,
+                variant: 'danger'
+            });
+        }
+    };
+
+    const copySavePathToClipboard = () => {
+        navigator.clipboard.writeText(`${savePathAdornmentText}${problemEditorForm.getValues().userPath}`);
+    };
+
+    const onFormChange = () => {
+        setAlertState({
+            message: '',
+            variant: 'info'
+        });
+    };
+
+    useEffect(() => {
+        (async () => {
+            const loadPromise = load();
+            const catalogPromise = getCatalog();
+            await Promise.all([loadPromise, catalogPromise]);    
```

Comment:
Why are you awaiting a promises here? Do they not complete if you don't wait for them?

Reply:
They do, I added this for supporting the loading spinner but never got to the loading spinner and it didn't seem to need it, in the future we should probably throttle the website and see what it looks like when requests aren't so quick